proposals/11-21-krm-fn-k8s-backend.md

```diff
@@ -0,0 +1,218 @@
+
+## Proposal
+
+The basis of this proposal is to give ability for the user to select that the krm-function would be ran
+*as-a-pod* rather than using `docker` (which will still be a default behavior that will guarantee no braking changes).
+For that purpose, in addition to the existing `--enable-exec`, `--enable-starlark` flags that have a very similar
+purpose - to allow the user *explicitly* say that it's ok to run krm-function not only in docker, we propose to add
+`--enable-as-a-pod` flag that will switch backend from `docker` to `kubectl`.
+
+In case the user has several k8s clusters, it's necessary to give him the ability to select
+where the krm-container will run. Kubectl has a pretty big list of parameters that allow
+to set kubeconfig location, context, namespace and etc. Adding all of them would be troublesome, but
+since we're only passing all of them to kubectl it's much easier to introduce only one parameter,
+`--pod-kubectl-args` that will contain all that arguments as one string, e.g.
+`--context kind-kustomize-api-test -n special-namespace`. This will give the user the maximum
+flexibility without limiting him. Even if kubectl will add some new parameters in future they all automatically
+will be supported.
+
+The current `docker` based implementation runs container in as-hermetic-as-possible environment:
+
+1. only env variables specified in function config OR env variables explicitly listed by the user are passed to krm-container.
+1. if krm-function requires network access it can be enabled only by explicitly set `--network` flag.
+1. if krm-function needs some additional mounted file/directory this also can be done only via setting a special `--mount` flag explicitly.
+
+That means that all non-default non-hermetic features except env variables are restricted and can be set only explicitly.
+Similarly, the restrictions applied to the k8s pod has to be set *explicitly* by the user with the same exception of env variables.
+
+Unfortunately, it's not possible to completely block network for a pod - networking is a k8s crucial part. But if needed
+it's possible to apply NetworkPolicies to the pod that match to it by labels. Since NetworkPolicy is a standard method
+to restrict pods, the user can apply one or more of them to the pods that run krm-function.
+NetworkPolicies can be applied only to the pods that match the label or expression that consists of set of labels.
+That means we have to give ability to user to set an arbitrary label(s) for the created pod.
+
+To make pod-based implementation compatible with major number of already existing set of krm-functions
+it's necessary for some of them to allow mounted files/directories if the user explicitly sets that.
+In contrast to the 'docker'-backend use-cases where user shares with krm a local file or directory,
+k8s doesn't allow to do this. That means that some krm-functions that require mounted files/directories
```

Comment:
One thing we can explore is that we can convert the files in the directory to a `ConfigMap` and then mount it to the pod for consumption. It can help with the use case when running from the client host.

Reply:
+1
This will work to emulate Read-only file mounts. ConfigMap has binaryData that will work well even for binary files if needed. We'll be able to emulate dir mounts as well just by creating configMaps for all internal files.
There won't be a way to get the the updated content from krm-function for read-write case. In practice I had couple of cases where I needed RW mount, so this is a pretty rare case I guess.

Reply:
I've updated this doc with description of that possible approach with ConfigMap. see line 88